src/main/java/com/google/devtools/build/lib/syntax/SkylarkList.java

```diff
@@ -201,6 +206,328 @@ public int hashCode() {
     return MutableList.copyOf(Mutability.IMMUTABLE, contents);
   }
 
+  /**
+   * A sequence returned by the {@code range} function invocation.
+   *
+   * <p>Instead of eagerly allocating an array with all elements of the sequence, this class uses
+   * simple math to compute a value at each index. This is particularly useful when range is huge or
+   * only a few elements from it are used.
+   *
+   * <p>Eventually {@code range} function should produce an instance of the {@code range} type as is
+   * the case in Python 3, but for now to preserve backwards compatibility with Python 2, {@code
+   * list} is returned.
+   */
+  @SkylarkModule(
+      name = "range",
+      category = SkylarkModuleCategory.BUILTIN,
+      doc =
+          "A language built-in type to support ranges. Example of range literal:<br>"
+              + "<pre class=language-python>x = range(1, 10, 3)</pre>"
+              + "Accessing elements is possible using indexing (starts from <code>0</code>):<br>"
+              + "<pre class=language-python>e = x[1]   # e == 2</pre>"
+              + "Ranges do not support the <code>+</code> operator for concatenation."
+              + "Similar to strings, ranges support slice operations:"
+              + "<pre class=language-python>range(10)[1:3]   # range(1, 3)\n"
+              + "range(10)[::2]  # range(0, 10, 2)\n"
+              + "range(10)[3:0:-1]  # range(3, 0, -1)</pre>"
+              + "Ranges are immutable, as in Python 3.")
+  public static final class RangeList extends SkylarkList<Integer> {
+
+    private final int step;
+    private final int start;
+
+    private static int computeItem(int start, int step, int index) {
+      return start + step * index;
+    }
+
+    /** Provides access to range elements based on their index. */
+    private static class RangeListView extends AbstractList<Integer> {
+
+      /** Iterator for increasing sequences. */
+      private static class IncreasingIterator extends UnmodifiableIterator<Integer> {
```

Comment:
In principle I thought it was fine that we were separating IncreasingIterator and DecreasingIterator for clarity purposes, but then it turns out that these classes are identical except for hasNext().

Perhaps we should combine them, and hasNext() can be   return (step > 0) ? : cursor < stop : cursor > stop;

Reply:
that's how I started, but then started prematurely optimizing to avoid extra branches :)